<a href="https://colab.research.google.com/github/Aseem2004/Concrete-Compressive-Strength-Analysis/blob/main/Concrete_Compressive_Strength.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Prediction of Concrete Compressive Strength using Linear Regression**

In [2]:
# Step 1: Import Library
import pandas as pd

In [3]:
# Step 2: Import Data
concrete=pd.read_csv('https://github.com/ybifoundation/Dataset/raw/main/Concrete%20Compressive%20Strength.csv')

In [4]:
concrete.head()

Cement (kg in a m^3 mixture)  Blast Furnace Slag (kg in a m^3 mixture)  \
0                         540.0                                       0.0   
1                         540.0                                       0.0   
2                         332.5                                     142.5   
3                         332.5                                     142.5   
4                         198.6                                     132.4   

   Fly Ash (kg in a m^3 mixture)  Water (kg in a m^3 mixture)  \
0                            0.0                        162.0   
1                            0.0                        162.0   
2                            0.0                        228.0   
3                            0.0                        228.0   
4                            0.0                        192.0   

   Superplasticizer (kg in a m^3 mixture)  \
0                                     2.5   
1                                     2.5   
2                                     0.0   
3                                     0.0   
4                                     0.0   

   Coarse Aggregate (kg in a m^3 mixture)  \
0                                  1040.0   
1                                  1055.0   
2                                   932.0   
3                                   932.0   
4                                   978.4   

   Fine Aggregate (kg in a m^3 mixture)  Age (day)  \
0                                 676.0         28   
1                                 676.0         28   
2                                 594.0        270   
3                                 594.0        365   
4                                 825.5        360   

   Concrete Compressive Strength(MPa, megapascals)   
0                                         79.986111  
1                                         61.887366  
2                                         40.269535  
3                                         41.052780  
4                                         44.296075

In [5]:
concrete.tail()

Cement (kg in a m^3 mixture)  Blast Furnace Slag (kg in a m^3 mixture)  \
1025                         276.4                                     116.0   
1026                         322.2                                       0.0   
1027                         148.5                                     139.4   
1028                         159.1                                     186.7   
1029                         260.9                                     100.5   

      Fly Ash (kg in a m^3 mixture)  Water (kg in a m^3 mixture)  \
1025                           90.3                        179.6   
1026                          115.6                        196.0   
1027                          108.6                        192.7   
1028                            0.0                        175.6   
1029                           78.3                        200.6   

      Superplasticizer (kg in a m^3 mixture)  \
1025                                     8.9   
1026                                    10.4   
1027                                     6.1   
1028                                    11.3   
1029                                     8.6   

      Coarse Aggregate (kg in a m^3 mixture)  \
1025                                   870.1   
1026                                   817.9   
1027                                   892.4   
1028                                   989.6   
1029                                   864.5   

      Fine Aggregate (kg in a m^3 mixture)  Age (day)  \
1025                                 768.3         28   
1026                                 813.4         28   
1027                                 780.0         28   
1028                                 788.9         28   
1029                                 761.5         28   

      Concrete Compressive Strength(MPa, megapascals)   
1025                                         44.284354  
1026                                         31.178794  
1027                                         23.696601  
1028                                         32.768036  
1029                                         32.401235

In [6]:
concrete.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   Cement (kg in a m^3 mixture)                      1030 non-null   float64
 1   Blast Furnace Slag (kg in a m^3 mixture)          1030 non-null   float64
 2   Fly Ash (kg in a m^3 mixture)                     1030 non-null   float64
 3   Water (kg in a m^3 mixture)                       1030 non-null   float64
 4   Superplasticizer (kg in a m^3 mixture)            1030 non-null   float64
 5   Coarse Aggregate (kg in a m^3 mixture)            1030 non-null   float64
 6   Fine Aggregate (kg in a m^3 mixture)              1030 non-null   float64
 7   Age (day)                                         1030 non-null   int64  
 8   Concrete Compressive Strength(MPa, megapascals)   1030 non-null   float64
dtypes: float64(8), int64

In [7]:
concrete.describe()

Cement (kg in a m^3 mixture)  Blast Furnace Slag (kg in a m^3 mixture)  \
count                   1030.000000                               1030.000000   
mean                     281.165631                                 73.895485   
std                      104.507142                                 86.279104   
min                      102.000000                                  0.000000   
25%                      192.375000                                  0.000000   
50%                      272.900000                                 22.000000   
75%                      350.000000                                142.950000   
max                      540.000000                                359.400000   

       Fly Ash (kg in a m^3 mixture)  Water (kg in a m^3 mixture)  \
count                    1030.000000                  1030.000000   
mean                       54.187136                   181.566359   
std                        63.996469                    21.355567   
min                         0.000000                   121.750000   
25%                         0.000000                   164.900000   
50%                         0.000000                   185.000000   
75%                       118.270000                   192.000000   
max                       200.100000                   247.000000   

       Superplasticizer (kg in a m^3 mixture)  \
count                             1030.000000   
mean                                 6.203112   
std                                  5.973492   
min                                  0.000000   
25%                                  0.000000   
50%                                  6.350000   
75%                                 10.160000   
max                                 32.200000   

       Coarse Aggregate (kg in a m^3 mixture)  \
count                             1030.000000   
mean                               972.918592   
std                                 77.753818   
min                                801.000000   
25%                                932.000000   
50%                                968.000000   
75%                               1029.400000   
max                               1145.000000   

       Fine Aggregate (kg in a m^3 mixture)    Age (day)  \
count                           1030.000000  1030.000000   
mean                             773.578883    45.662136   
std                               80.175427    63.169912   
min                              594.000000     1.000000   
25%                              730.950000     7.000000   
50%                              779.510000    28.000000   
75%                              824.000000    56.000000   
max                              992.600000   365.000000   

       Concrete Compressive Strength(MPa, megapascals)   
count                                       1030.000000  
mean                                          35.817836  
std                                           16.705679  
min                                            2.331808  
25%                                           23.707115  
50%                                           34.442774  
75%                                           46.136287  
max                                           82.599225

In [8]:
concrete.shape

(1030, 9)

In [9]:
concrete.columns

Index(['Cement (kg in a m^3 mixture)',
       'Blast Furnace Slag (kg in a m^3 mixture)',
       'Fly Ash (kg in a m^3 mixture)', 'Water (kg in a m^3 mixture)',
       'Superplasticizer (kg in a m^3 mixture)',
       'Coarse Aggregate (kg in a m^3 mixture)',
       'Fine Aggregate (kg in a m^3 mixture)', 'Age (day)',
       'Concrete Compressive Strength(MPa, megapascals) '],
      dtype='object')

In [10]:
# Step 3: Define x and y
y=concrete['Concrete Compressive Strength(MPa, megapascals) ']

In [11]:
x=concrete[['Cement (kg in a m^3 mixture)',
'Blast Furnace Slag (kg in a m^3 mixture)',
'Fly Ash (kg in a m^3 mixture)', 'Water (kg in a m^3 mixture)',
'Superplasticizer (kg in a m^3 mixture)',
'Coarse Aggregate (kg in a m^3 mixture)',
'Fine Aggregate (kg in a m^3 mixture)', 'Age (day)']]

In [12]:
# Step 4: train_test_split
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,train_size=0.7,random_state=42)

In [13]:
x_train.shape,x_test.shape,y_train.shape,y_test.shape

((721, 8), (309, 8), (721,), (309,))

In [14]:
# Step 5: Choose the model
from sklearn.linear_model import LinearRegression
model=LinearRegression()

In [15]:
# Step 6: Train the model
model.fit(x_train,y_train)

LinearRegression()

In [16]:
model.intercept_

-59.45903971993882

In [17]:
model.coef_

array([ 0.12544516,  0.11677346,  0.09002765, -0.09095886,  0.39490648,
        0.02805813,  0.0363444 ,  0.11394849])

Clearly, the equation of best fit here is y=59.45903971993882 + (0.12544516)a + (0.11677346)b + (0.09002765)c + (-0.09095886)d + (0.39490648)e  = (0.02805813)f + (0.0363444)g + (0.11394849)h

where a,b,c,d,e,f,g and h are the parameters i.e. 'Cement (kg in a m^3 mixture)',
'Blast Furnace Slag (kg in a m^3 mixture)',
'Fly Ash (kg in a m^3 mixture)', 'Water (kg in a m^3 mixture)',
'Superplasticizer (kg in a m^3 mixture)',
'Coarse Aggregate (kg in a m^3 mixture)',
'Fine Aggregate (kg in a m^3 mixture)', 'Age (day)'

In [18]:
# Step 7: Predict with the model
y_pred=model.predict(x_test)

In [19]:
y_pred

array([58.57504735, 52.45260803, 64.10542477, 51.99681408, 17.52475822,
       39.59501382, 26.65823233, 45.03359831, 30.0486104 , 37.57580535,
       27.89765125, 19.09784563, 68.04009097, 52.75209712, 30.01212472,
       44.32222404, 29.17851117, 26.34893715, 32.15306403, 32.35591276,
       36.66984261, 31.61981735, 38.55399371, 25.05929092, 32.81848049,
       34.56866584, 14.60378741, 40.05211442, 41.86352172, 21.0350626 ,
       35.30346703, 30.51099918, 43.39339578, 45.46205927, 31.11058479,
       29.08237973, 29.49268523, 38.62442885, 20.43472484, 38.7226507 ,
       21.29200171, 16.11500654, 31.16081174, 51.48092385, 20.88482357,
       57.93499638, 51.04728239, 59.76068149, 20.17329791, 19.32472416,
       40.38374696, 35.78563857, 29.86134962, 33.48866724, 46.54284745,
       51.8957266 , 28.45333159, 15.78997941, 30.12624922, 18.20968721,
       38.37882512, 19.58827592, 31.93782676, 56.11230154, 22.88593483,
       21.31701385, 32.44678829, 16.79579322, 26.11432518, 26.07

In [20]:
# Step 8: Accuracy of model
from sklearn.metrics import mean_absolute_error,mean_absolute_percentage_error,mean_squared_error

In [21]:
mean_absolute_error(y_test,y_pred)

8.298641626746605

In [22]:
mean_squared_error(y_test,y_pred)

109.75614063771856

In [23]:
mean_absolute_percentage_error(y_test,y_pred)

0.3219092251031804

This means that error in prediction is 32.19% approx and so our model is 100-32.19=67.81% accurate.